src/server/pps/server/api_server.go

```diff
@@ -1531,9 +1534,37 @@ func (a *apiServer) getLogsFromStats(pachClient *client.APIClient, request *pps.
 	return eg.Wait()
 }
 
+func (a *apiServer) validatePipelineRequest(request *pps.CreatePipelineRequest) error {
+	if request.Pipeline == nil {
+		return goerr.New("invalid pipeline spec: request.Pipeline cannot be nil")
+	}
+	if request.Pipeline.Name == "" {
+		return goerr.New("invalid pipeline spec: request.Pipeline.Name cannot be empty")
+	}
```

Comment:
It looks like some of this validation overlaps with the logic in `validatePipeline` is there a good reason to have it in both places or can it be consolidated?

Reply:
Perhaps—the reason I created a separate function has to do with when the two are called. `validatePipelineRequest` takes a `CreatePipelineRequest` and is called before the `PipelineInfo` is create and `setPipelineDefaults` is called, and `validatePipeline` takes a `PipelineInfo` and is called after `setPipelineDefaults`.

Certain fields, like `request.Pipeline` and `request.Pipeline.Transform`, will cause `setPipelineDefaults` or the initialization of the `PipelineInfo` to panic, so it makes sense to validate them as early as possible. But It also makes sense to validate the `PipelineInfo` as late as possible, before we write it, to make sure `setPipelineDefaults` and the surrounding logic doesn't accidentally give us an invalid struct to persist.

At least, that was the rationale. If you think this should be smaller I can pull out all the checks that are the same, but I do think there needs to be to be at least a few pieces of validation logic here at the front.

Reply:
I see, I think that's a reasonable argument for having 2 separate pieces of validation logic.